tests/periph_rtt_min: add test to figure out RTT_MIN_OFFSET

fjmolinas opened the pull request:
### Contribution description

This PR adds a simple test to figure out the correct value for `RTT_MIN_VALUE`

IMO this is also a good reason to make `RTT_MIN_VALUE` public and not only in `ztimer` code.

### Testing procedure

`BOARD=<board>make -C tests/periph_rtt_min/ flash test`

```
main(): This is RIOT! (Version: 2020.07-devel-774-g243984-pr_rtt_min_value_test)
Evaluate RTT_MIN_VALUE over 1024 samples
...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
RTT_MIN_VALUE for samr21-xpro: 10
```

```
main(): This is RIOT! (Version: 2020.07-devel-774-g243984-pr_rtt_min_value_test)
Evaluate RTT_MIN_VALUE over 1024 samples
...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
RTT_MIN_VALUE for iotlab-m3: 2

```


You can also see that if in `tests/periph_rtt` `TICKS_TO_WAIT` is set lower than `RTT_MIN_VALUE` the test does not work.

Note hardware specification may suggest a value different than the one resolved by the test, e.g. `cc2538` points out a value of 5. It will also depend on the `RTT_FREQUENCY`, should be ran at the highest possible frequency.

### Issues/PRs references

Figure out while debugging why changing to `ztimer` broke support for `samr21-xpro` in https://github.com/RIOT-OS/RIOT/pull/13824.

aabadie commented on tests/periph_rtt_min/README.md:
```diff
@@ -0,0 +1,7 @@
+## About
+
+Describe here what the test application is about
```
Maybe "Describe here what the test application is about" ? :)
(and thanks for using `make generate-test` btw ;))

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,74 @@
+#include "timex.h"
+#include "ztimer.h"
+
+#define MIN_SLEEP       (100)
```
```suggestion
#define MIN_SLEEP_US    (100)
```

fjmolinas commented:
Rebased otherwise there would be conflict due to the `stm32` changes.

benpicco commented on tests/periph_rtt_min/README.md:
```diff
@@ -0,0 +1,18 @@
+## About
+
+This applications is meant to determine RTT_MIN_VALUE for a specific BOARD.
```
What is `RTT_MIN_VALUE` used for? What's it's meaning?

fjmolinas replied:
I want to change it to `RTT_MIN_OFFSET` https://github.com/RIOT-OS/RIOT/pull/14259, currently this value is only used in `ztimer` to avoid setting callbacks that wont be triggered.

benpicco replied:
Better add a comment.

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,75 @@
+
+#include "xtimer.h"
+
+#define MAX_WAIT_US     (100)
```
Why is this needed?

fjmolinas replied:
I wanted to wait for a minimum amount of time to account for the time between the callback being set and handled.

benpicco commented:
Please squash!

aabadie commented:
Please rebase and adapt the name to RTT_MIN_OFFSET instead of RTT_MIN_VALUE.

fjmolinas commented:
Rebased.

fjmolinas commented:
Should I change the name to `tests/periph_rtt_min_offset`?

kaspar030 commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,76 @@
+    for (unsigned i = 0; i < SAMPLES; i++) {
+        uint32_t offset = 0;
+        cb_triggered = false;
+        while (cb_triggered == false) {
```
I'm trying to get this logic. basically it is setting an alarm, waiting the time it should take using xtimer, then verifying if the alarm triggered. IMO that's a bit brittle.
For synchronized RTTs, that'll probably work, but if the RTT works like a regular timer and just quantizes on the RTT clock, every set is late by `0<x<1` ticks, as the set_alarm() can happen anywhere within one RTT clock cycle.
This is supposed to figure out the smallest value that can be set without underflowing, right?
Would it maybe make sense to use the same logic as done for periph_timer_short_relative_set?
There, values are tried from larger to 0, until the timer clearly doesn't trigger anymore.

fjmolinas replied:
In the case of `RTT` trying from larger values down might mean waiting for a long time, some of our RTT are running at `1Hz`. And why would going down from larger be better than going up? Seems just as brittle to me.

fjmolinas replied:
ping @kaspar030!

fjmolinas replied:
Ping @kaspar030?

benpicco replied:
How about using `xtimer_mutex_lock_timeout()` instead?

fjmolinas replied:
Did that, also found a bug in stm32f1 rtt code.

fjmolinas replied:
Will split the stm32 change out.

fjmolinas replied:
> I'm trying to get this logic. basically it is setting an alarm, waiting the time it should take using xtimer, then verifying if the alarm triggered. IMO that's a bit brittle.

It is waiting always longer then the time it should take to be set, I'm always waiting at least 100ul which is about 3 ticks for a 32Khz timer, maybe I can set that too ticks related.

fjmolinas replied:
I added a 3ticks margin to the time waited.

kaspar030 replied:
discussed offline, no issue here from my side anymore.

fjmolinas commented:
Split the stm32 fix out

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,78 @@
+
+void cb(void *arg)
+{
+    (void)arg;
+    mutex_unlock(&_mtx);
```
```suggestion
    mutex_unlock(arg);
```

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,78 @@
+#define SAMPLES 1024LU
+#endif
+
+static mutex_t _mtx = MUTEX_INIT_LOCKED;
+
```
```suggestion
```

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,78 @@
+
+int main(void)
+{
+    uint32_t value = 0;
```
```suggestion
    uint32_t value = 0;
    mutex_t lock = MUTEX_INIT_LOCKED;
```

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,78 @@
+            offset++;
+            uint32_t now = rtt_get_counter();
+            mutex_trylock(&_mtx);
+            rtt_set_alarm((now + offset) % RTT_MAX_VALUE, cb, 0);
```
```suggestion
            rtt_set_alarm((now + offset) % RTT_MAX_VALUE, cb, &lock);
```

benpicco commented on tests/periph_rtt_min/main.c:
```diff
@@ -0,0 +1,78 @@
+        while (ret != 0) {
+            offset++;
+            uint32_t now = rtt_get_counter();
+            mutex_trylock(&_mtx);
```
What is that for?

fjmolinas replied:
Since there are multiple iterations you always want the mutex to be locked before `xtimer_mutex_lock_timeout`, if it timed out then it would not be locked on each new "iteration". I added a comment.

benpicco replied:
> if it timed out then it would not be locked on each new "iteration"

If `xtimer_mutex_lock_timeout()` times out the mutex is still locked.

fjmolinas replied:
True I'll remove.

fjmolinas commented:
@benpicco I addressed your comments.

benpicco commented:
Please squash!

fjmolinas commented:
@benpicco I added bbee446c08333e2e4420284690ae5ad82fd0ac18 to add some more details on the application results as well as wrap the `rtt` calls in a `irq_enable` `irq_disable` block

benpicco commented:
With disabling IRQs we won't be able to spot bugs like #14779
The RTT API should be safe to use without disabling IRQs :wink:

fjmolinas commented:
> With disabling IRQs we won't be able to spot bugs like #14779
> The RTT API should be safe to use without disabling IRQs wink

Removed and squashed.

fjmolinas commented:
All green here @benpicco @kaspar030

benpicco reviewed (approved):
I just gave it another try, works fine and the code looks good.
